sound/soc/sof/sof-audio.c

```diff
@@ -466,48 +467,99 @@ struct snd_sof_dai *snd_sof_find_dai(struct snd_soc_component *scomp,
 /*
  * SOF Driver enumeration.
  */
-static int sof_machine_check(struct snd_sof_dev *sdev)
+static int sof_machine_check(struct platform_device *pdev,
+			     const struct sof_dev_desc *desc)
 {
-	struct snd_sof_pdata *plat_data = sdev->pdata;
+	struct sof_audio_dev *sof_audio = sof_mfd_get_client_data(&pdev->dev);
 #if IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)
 	struct snd_soc_acpi_mach *machine;
 	int ret;
 #endif
 
-	if (plat_data->machine)
+	if (sof_audio->machine)
 		return 0;
 
 #if !IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)
-	dev_err(sdev->dev, "error: no matching ASoC machine driver found - aborting probe\n");
+	dev_warn(&pdev->dev, "error: no matching ASoC machine driver found - aborting probe\n");
 	return -ENODEV;
 #else
 	/* fallback to nocodec mode */
-	dev_warn(sdev->dev, "No ASoC machine driver found - using nocodec\n");
-	machine = devm_kzalloc(sdev->dev, sizeof(*machine), GFP_KERNEL);
+	dev_warn(&pdev->dev, "No ASoC machine driver found - using nocodec\n");
+	machine = devm_kzalloc(&pdev->dev, sizeof(*machine), GFP_KERNEL);
 	if (!machine)
 		return -ENOMEM;
 
-	ret = sof_nocodec_setup(sdev->dev, plat_data, machine,
-				plat_data->desc, plat_data->desc->ops);
+	ret = sof_nocodec_setup(&pdev->dev, sof_audio, machine, desc);
 	if (ret < 0)
 		return ret;
 
-	plat_data->machine = machine;
+	machine->mach_params.platform = dev_name(&pdev->dev);
+
+	sof_audio->machine = machine;
 
 	return 0;
 #endif
 }
 
+static int sof_audio_select_machine(struct platform_device *pdev,
+				    const struct sof_dev_desc *desc)
+{
+	struct sof_audio_dev *sof_audio = sof_mfd_get_client_data(&pdev->dev);
+	struct snd_soc_acpi_mach *mach;
+	int ret;
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_FORCE_NOCODEC_MODE)
+	/* force nocodec mode */
+	dev_warn(&pdev->dev, "Force to use nocodec mode\n");
+	mach = devm_kzalloc(&pdev->dev, sizeof(*mach), GFP_KERNEL);
+	if (!mach)
+		return -ENOMEM;
+
+	ret = sof_nocodec_setup(&pdev->dev, sof_audio, mach, desc);
+	if (ret < 0)
+		return ret;
+	mach->mach_params.platform = dev_name(&pdev->dev);
+#else
+	/* find machine */
+	mach = snd_soc_acpi_find_machine(desc->machines);
+	if (!mach) {
```

Comment:
This function is not generic enough with respect to ACPI / DT implementation. This can be generic for Intel platforms. We can leave it like now and fix it later for DT.

One thing could be fixed now I think. In the no-codec case the previous implementation did not set mach->mach_params.platform, so I could also use the sample implementation for nocodec on ARM.

In this new patch nocodec scenario touches mach_params.platform. Can you somehow avoid this?

Reply:
@dbaluta with the latest changes all the acpi match stuff goes into the intel-specific audio clients. I suppose this would completely break your platform. I'll have to work with you to make this work on iMX. I am traveling the next couple of days but I can maybe get in touch with you on Monday?

Reply:
@ranj063 Monday is fine for me. Lets have a talk on Monday morning/afternoon your time.